File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        n = len(nums)
        left, right = 0, n - 1
        res = [-1, -1]
        #find the left most
        while(left <= right):
            mid = left + (right - left) // 2
            if nums[mid] == target:
                if mid == 0 or nums[mid-1] != target:
                    res = [mid, mid]
                    break
                else:
                    right = mid - 1
            elif nums[mid] > target:
                right = mid - 1
            else:
                left = mid + 1
        
        #find the rightmost
        if res[0] == -1:
            return [-1,-1]
        start, end = res[0], n - 1
        while(start < end and nums[start + 1] == target):
            start += 1
        return [res[0], start]
            

File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        //find the leftmost
        int leftMost = -1, rightMost = -1;
        while(left <= right){
            int mid = left + (right - left) / 2;
            if(nums[mid] == target){
                if(mid == 0 || (nums[mid-1] != target)){
                    leftMost = mid;
                    break;
                }else{
                    right = mid - 1;
                }
            }else if(nums[mid] > target){
                right = mid - 1;
            }else{
                left = mid + 1;
            }
        }
        
        left = 0; right = nums.length - 1;
        while(left <= right){
            int mid = left + (right - left) / 2;
            if(nums[mid] == target){
                if(mid == nums.length - 1 || (nums[mid+1] != target)){
                    rightMost = mid;
                    break;
                }else{
                    left = mid + 1;
                }
            }else if(nums[mid] > target){
                right = mid - 1;
            }else{
                left = mid + 1;
            }
        }
        return new int[]{leftMost, rightMost};
        
        
        
    }
}